### DISCUSSION TITLE
tasmota.set_timer not working in method of module's class

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#18004] [Q&A]
I have a simple class for ports, sitwch them on, off or on+delayed_off.

testing `fill_water(2000)` in the berry cosole works fine.

i have defined the class in the file water_module.be and load this file as module in web.be.

`web.be` ist for generating the webbutton.
`water_module.be` has the methods inside.

If starting `fill_water(2000)` by pressing the webbutton the function waterpump.set() failed. It hangs on
`tasmota.set_timer(t, self.set)`
i got the following console output:
`--> BRY: Exception> 'type_error' - BRY: method not allowed, use a closure like '/ args -> obj.func(args)'`
Why or how can i workaround?

web.be
```
import webserver # import webserver class
import water_module

class MyButtonMethods : Driver

  #- create a method for adding a button to the main menu -#
  def web_add_main_button()
    webserver.content_send("<p></p><button onclick='la(\"&m_toggle_button=1\");'>Start fill water</button>")
  end

  #- As we can add only one sensor method we will have to combine them besides all other sensor readings in one method -#
  def web_sensor()
  
    if webserver.has_arg("m_toggle_button")
      print("button pressed")
      bewasserung_module.FillWater()
    end

  end
end

d1 = MyButtonMethods()
tasmota.add_driver(d1)
```

water_module.be
```
var water_module = module("water_module")    # use `import water_module`

#==================================#
#====== Globale Variablen   =======#

var TIME_FILL_WATER = 20000 #20sek

var PORT_PUMP_WATER= 3
var NAME_PUMP_WATER="waterpump"


#==================================#
#=========== FUNCTIONS   ==========#

class pumps

    var name, portnumber

    def init(name, portnumber)
        self.name=name
        self.portnumber=portnumber
    end

    def set(t)     # t=0 dauer an, t=nil aus, t>1 tiemout in ms

        print("Pump:  ", self.name, "  Portnumber:  ", self.portnumber)

        if t==nil
            print(" OFF:  ")
            tasmota.set_power(self.portnumber, false) 

        elif t==0
            print(" ON:  ")
            tasmota.set_power(self.portnumber, true) 

        else
            print(" ON for:  ", t, "ms")
            tasmota.set_power(self.portnumber, true) 
        
            # start timer for shutt off
            tasmota.set_timer(t, self.set)                    # --> BRY: Exception> 'type_error' - BRY: method not allowed, use a closure like '/ args -> obj.func(args)'

        end
    end
 end


var waterpump = pumps(NAME_PUMP_WATER, PORT_PUMP_WATER)


#----------- Fill Water--------------#
def fill_water()
    tasmota.set_timer(TIME_FILL_WATER,waterpump.set)     # on for TIME_FILL_WATER ms
end


#==================================#
#========  Module FUNCTIONS  ======#
# impordet by web.be 

# Fill water - called if web button is puhesd
water_module.FillWater = def ()
    fill_water(0)

end


print("Module loaded")
return water_module    # return the module as the output of import
```

thx alot!

**Comment by sfromis** [comment]
`BRY: Exception> 'type_error' - BRY: method not allowed, use a closure like '/ args -> obj.func(args)'`

The reason for this is that the function called by `tasmota.set_timer` is a "free-floading" function outside the context of the class, meaning that it does not have `self` or instance variables. This is not related to it being inside a module.

What you can do is to have a static method not needing access to the instance (or `self`) which may not be practical, or just follow the very specific recommendation detailed in the error message. Coding it as a closure means that the code inside the anonymous function created does inherit access to the local context (both local vars, instance and `self`), and can pass that on to any functions called.

**Comment by ChrisVonFrosta** [comment]
Thank you for your support!!
I'm not familiar with the concept of a closure.  So, can I rewrite the class method set(t) as a closure? I have looked at some closure examples in Python, but I couldn't adapt them to my code. Could you please provide an example related to my class or a stronger hint? :)

I suspect that this might have also been the cause of the problem with the global variables in the [discussion,](https://github.com/arendst/Tasmota/discussions/17977#discussioncomment-5029502) right?"

**Reply by sfromis** [reply]
Again, just do as suggested in the error message. `use a closure like '/ args -> obj.func(args)`
In your case, `/->self.set()` instead of just `self.set` as your `obj` is the current instance, `self` and you have no arguments.

That notation is short form for `def () self.set() end` which is an anonymous function definition having access to the context from where the code is placed (this is what makes it a closure), meaning that it can call `set()` as an instance method. It would also be possible to code the whole `set` function directly there, instead of creating a function to call the set function, but that may not be practical always.

**Comment by ChrisVonFrosta** [comment]
I'm afraid to ask any further. I've read your instructions several times and made countless attempts, but unfortunately, I don't get it. Of course, I also googled it.

However, with every attempt, I still receive the error message:
`BRY: Exception> 'type_error' - BRY: method not allowed, use a closure like '/ args -> obj.func(args)'`

Apparently, I don't understand it correctly. I have currently modified the class as follows:
```
class pumps

    var name, portnumber

    def init(name, portnumber)
        self.name=name
        self.portnumber=portnumber
    end

    def set(t)     # t=0 dauer an, t=nil aus, t>1 tiemout in ms

        print("Pump:  ", self.name, "  Portnumber:  ", self.portnumber)

        if t==nil
            print(" OFF:  ")
            tasmota.set_power(self.portnumber, false) 

        elif t==0
            print(" ON:  ")
            tasmota.set_power(self.portnumber, true) 

        else
            print(" ON for:  ", t, "ms")
            tasmota.set_power(self.portnumber, true) 
        
            # start timer for shutt off
            tasmota.set_timer(t, /->self.set)                    # --> BRY: Exception> 'type_error' - BRY: method not allowed, use a closure like '/ args -> obj.func(args)'

        end
    end
 end
```

**Reply by sfromis** [reply]
Read more carefully then. I told you to use `/->self.set()` and you used `/->self.set` instead. And you did not do it all places where you have `tasmota.set_timer` to invoke the `set` instance method in the class, depending on how much you did not post, at least also in that `fill_water` function. Of course, using `self` is only inside the class, outside of it you'll have to use the variable where you have the class instance.

**Comment by ChrisVonFrosta** [comment]
I had also tried "/->set()" but it didn't work. However, you are also right that I had copied only snippets of my code. Now, I have created a new "clean project" just for this case, and voilà, it works... thank you very much!

[water_module.be.txt](https://github.com/arendst/Tasmota/files/10794998/water_module.be.txt)
[web.be.txt](https://github.com/arendst/Tasmota/files/10794999/web.be.txt)
